pex/platforms.py

```diff
@@ -108,41 +92,34 @@ def from_tag(cls, tag):
         impl, version = tag.interpreter[:2], tag.interpreter[2:]
         return cls(platform=tag.platform, impl=impl, version=version, abi=tag.abi)
 
-    def __new__(cls, platform, impl, version, abi):
-        if not all((platform, impl, version, abi)):
-            raise cls.InvalidPlatformError(
-                "Platform specifiers cannot have blank fields. Given platform={platform!r}, "
-                "impl={impl!r}, version={version!r}, abi={abi!r}".format(
-                    platform=platform, impl=impl, version=version, abi=abi
+    platform = attr.ib(converter=_normalize_platform)  # type: str
+    impl = attr.ib()  # type: str
+    version = attr.ib()  # type: str
+    abi = attr.ib()  # type: str
+
+    @platform.validator
+    @impl.validator
+    @version.validator
+    @abi.validator
+    def _non_blank(self, attribute, value):
+        if not value:
+            raise self.InvalidPlatformError(
+                "Platform specifiers cannot have blank fields. Given {field}={value!r}".format(
+                    field=attribute.name, value=value
                 )
             )
-        platform = platform.replace("-", "_").replace(".", "_")
-        abi = cls._maybe_prefix_abi(impl, version, abi)
-        return super(Platform, cls).__new__(cls, platform, impl, version, abi)
-
-    @property
-    def platform(self):
-        # type: () -> str
-        return cast(str, super(Platform, self).platform)
-
-    @property
-    def impl(self):
-        # type: () -> str
-        return cast(str, super(Platform, self).impl)
-
-    @property
-    def version(self):
-        # type: () -> str
-        return cast(str, super(Platform, self).version)
 
-    @property
-    def abi(self):
-        # type: () -> str
-        return cast(str, super(Platform, self).abi)
+    def __attrs_post_init__(self):
+        # type: () -> None
+        if self.impl == "cp" and not self.abi.startswith(self.interpreter):
+            # N.B. This permits CPython users to pass in simpler extended platform
+            # strings like `linux-x86_64-cp-27-mu` vs e.g. `linux-x86_64-cp-27-cp27mu`.
+            object.__setattr__(self, "abi", self.interpreter + self.abi)
```

Comment:
N.B.: `object.__setattr__` in `__attrs_post_init__` is a native way (in their docs) to do frozen after init in attrs.

Reply:
This conversation can be resolved at your discretion.

It looks like you left this explicit comment https://github.com/pantsbuild/pex/pull/1234#issuecomment-780024126 pointing to the API reference for `attrs`, so it feels like your choice to avoid it in the code itself was intentional. `attrs` is canonical enough of a library that I think it's relatively straightforward to later readers to web search for `attrs` directly, which would allow them to understand the rationale for using this method. I would still have left a comment in the code saying:

```python
            # N.B.: object.__setattr__ in __attrs_post_init__ is a native way (in their docs at https://www.attrs.org/en/stable/api.html) to do frozen after init in attrs.
```

but I would not have any difficulty obtaining context later if you chose otherwise.

Reply:
Yeah - I view that sort of thing as babying readers. If you adopt a library you should not regurgitate its docs. You should learn the library. I'm a bit more OK with babying reviewers on the 1st intro of the library - thus the PR comment.